### PR TITLE
Ticket1222 instrument list refactor

### PR DESCRIPTION
**Pull request by AdrianPotter**
### Description of work

The InstrumentListObservable class was previously treated as a glorified collection rather than an observable. It now properly implements the observable pattern which has simplified it significantly. Responsibility for waiting for the initial instruments list has been shifted more appropriately to the instrument class. Eventually, we may be able to do without the wait altogether.

### To test

https://github.com/ISISComputingGroup/IBEX/issues/1222

### Acceptance criteria

- [x] InstrumentListObservable is now an observable
- [x] Behaviour of the GUI is unchanged noting in particular
    - [x] The initial instrument is the same as the instrument at last exit from the GUI
    - [x] The select instrument dialog has the correct instruments listed consistently

---

#### Code Review

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/GUI-Coding-Conventions)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit tests in place? Are the unit tests small and test the a class in isolation?
- [x] Are there [system tests](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/System-Testing-with-RCPTT) in place? Do they test a minimal set of functionality and leave the gui as close as possible to its original state?
- [x] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?

### Functional Tests

- [x] Do changes function as described? Add comments below that describe the tests performed.
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has developer documentation been updated if required?

### Final Steps
- [x] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

File: base/uk.ac.stfc.isis.ibex.instrument/src/uk/ac/stfc/isis/ibex/instrument/Instrument.java [hunk 1]
```diff
@@ -51,149 +52,282 @@
...
+            }
+
+            @Override
+            public void onError(Exception e) {
+                LOG.error("Error whilst reading instrument list : " + e.getMessage());
+            }
+
+            @Override
+            public void onConnectionStatus(boolean isConnected) {
+            }
```
Comment (John-Holt-Tessella) [inline base/uk.ac.stfc.isis.ibex.instrument/src/uk/ac/stfc/isis/ibex/instrument/Instrument.java]: Could you add an error message here please?

File: base/uk.ac.stfc.isis.ibex.instrument/src/uk/ac/stfc/isis/ibex/instrument/Instrument.java [hunk 2]
```diff
@@ -51,149 +52,282 @@
...
+    private static void updateExtendingPlugins(InstrumentInfo selectedInstrument) {
+        IExtensionRegistry registry = Platform.getExtensionRegistry();
+        IConfigurationElement[] elements = registry.getConfigurationElementsFor("uk.ac.stfc.isis.ibex.instrument.info");
+        for (IConfigurationElement element : elements) {
+            try {
+                final Object obj = element.createExecutableExtension("class");
+                InstrumentInfoReceiver receiver = (InstrumentInfoReceiver) obj;
+                receiver.setInstrument(selectedInstrument);
+            } catch (CoreException e) {
                 e.printStackTrace();
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.instrument/src/uk/ac/stfc/isis/ibex/instrument/Instrument.java]: :white_check_mark:

File: base/uk.ac.stfc.isis.ibex.instrument/src/uk/ac/stfc/isis/ibex/instrument/Instrument.java (hunk 1 above)
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.instrument/src/uk/ac/stfc/isis/ibex/instrument/Instrument.java]: :white_check_mark:

File: base/uk.ac.stfc.isis.ibex.instrument/src/uk/ac/stfc/isis/ibex/instrument/Instrument.java (hunk 2 above)
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.instrument/src/uk/ac/stfc/isis/ibex/instrument/Instrument.java]: :white_check_mark:

File: base/uk.ac.stfc.isis.ibex.instrument/src/uk/ac/stfc/isis/ibex/instrument/Instrument.java [hunk 3]
```diff
@@ -51,149 +52,282 @@
...
+     * Set the initial instrument which is used at the subsequent startup.
+     */
+    public void setInitial() {
+        initalPreference.put(initialInstrument, currentInstrument().name());
+
         try {
             // forces the application to save the preferences
             initalPreference.flush();
         } catch (BackingStoreException e2) {
             e2.printStackTrace();
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.instrument/src/uk/ac/stfc/isis/ibex/instrument/Instrument.java]: :white_check_mark:

**Comment by DominicOram** [comment]
Couple of requests:
* The error formatter looks like it could be useful throughout the UI, could you put it in the logger plugin? And add a ticket to use it in places in the GUI where it may be useful?
* Can you use the formatter on the onError in setInstrumentObserver?

**Comment by AdrianPotter** [comment]
Needed quite a messy merge with custom instrument names code from master. Make sure I haven't broken it!

**Pull request merged**